### ISSUE TITLE
Refresh token

### ISSUE THREAD

**Issue by henri-tremblay** [kink80/cognitoauth#2]
I am using your awesome code to authenticate with cognito. However, I am now struggling to refresh the token. It should be fairly straightforward (pass the refresh token, pass the device id and I'm done). But it fails.

Are you currently doing it? And if it's the case, could you please share this code?

Thanks.

**Comment by kink80** [comment]
Hi Henri, I'm glad that you liked this library. My main focus to get Cognito auth working so I intentionally left out dealing with refresh tokens.

However looking into the official SDK at https://github.com/aws/aws-sdk-android/blob/master/aws-android-sdk-cognitoauth/src/main/java/com/amazonaws/mobileconnectors/cognitoauth/AuthClient.java you need to issue refresh token request and update the session. 

Let me know how you succeeded with this approach and if you don't I'll try looking into it.

Thanks

**Comment by henri-tremblay** [comment]
So far I have lost my fight by getting and invalid token for everything I tried.

```
AuthenticationResultType authenticationResultType = ... retrieved using your framework

AWSCognitoIdentityProvider identityProviderClient = AWSCognitoIdentityProviderClientBuilder.standard()
    .withCredentials(new AWSStaticCredentialsProvider(new AnonymousAWSCredentials()))
    .withRegion(Regions.US_EAST_1)
    .build();

Map<String, String> params = new HashMap<>();
params.put("DEVICE_KEY", authenticationResultType.getNewDeviceMetadata().getDeviceKey());
params.put("REFRESH_TOKEN", authenticationResultType.getRefreshToken());

InitiateAuthRequest authRequest = new InitiateAuthRequest()
    .withAuthFlow(AuthFlowType.REFRESH_TOKEN_AUTH)
    .withClientId(appId)
    .withAuthParameters(params);

InitiateAuthResult authResult = identityProviderClient.initiateAuth(authRequest);
```
And this is what fails. I've tried multiple permutations of parameters. Still fail. So yes if you have an idea, it would be highly helpful.

**Comment by kink80** [comment]
Try using `AuthFlowType.REFRESH_TOKEN` instead, that works for me.

**Comment by henri-tremblay** [comment]
Sad. I was so hopeful. But still the same error

> com.amazonaws.services.cognitoidp.model.NotAuthorizedException: Invalid Refresh Token. (Service: AWSCognitoIdentityProvider; Status Code: 400; Error Code: NotAuthorizedException; Request ID: xxxx

Here is a full test if it helps. And thanks a lot for helping me. I was able to refresh using JavaScript. So I don't think it is a right issue. But it's getting weird.

```java
public class RefreshTest {

    private String appId = "xxxxx";
    private String poolId = "us-east-1_xxxxx";
    private String user = "demo@stuff.com";
    private String password = "password";

    @Test
    public void test() {
        AWSCognitoIdentityProvider identityProviderClient = AWSCognitoIdentityProviderClientBuilder.standard()
            .withCredentials(new AWSStaticCredentialsProvider(new AnonymousAWSCredentials()))
            .withRegion(Regions.US_EAST_1)
            .build();

        AWSCryptoSettings cryptoParams = new AWSCryptoSettings();
        AWSCognitoSession clientSession = new AWSCognitoSession(cryptoParams, user, password, poolId);

        InitiateAuthRequest authRequest = new InitiateAuthRequest()
            .withAuthFlow(AuthFlowType.USER_SRP_AUTH)
            .withClientId(appId)
            .withAuthParameters(clientSession.step1());

        InitiateAuthResult authResult = identityProviderClient.initiateAuth(authRequest);

        Map<String, String> params = authResult.getChallengeParameters();
        Map<String, String> srpAuthResponses = clientSession.step2(params);

        RespondToAuthChallengeRequest respondToAuthChallengeRequest = new RespondToAuthChallengeRequest()
            .withChallengeName(authResult.getChallengeName())
            .withClientId(appId)
            .withChallengeResponses(srpAuthResponses);

        RespondToAuthChallengeResult respondToAuthChallengeResult = identityProviderClient.respondToAuthChallenge(respondToAuthChallengeRequest);
        AuthenticationResultType authenticationResultType = respondToAuthChallengeResult.getAuthenticationResult();
        String deviceKey = authenticationResultType.getNewDeviceMetadata().getDeviceKey();
        String refreshToken = authenticationResultType.getRefreshToken();

        Map<String, String> refreshParams = new HashMap<>();
        refreshParams.put("DEVICE_KEY", deviceKey);
        refreshParams.put("REFRESH_TOKEN", refreshToken);

        authRequest = new InitiateAuthRequest()
            .withAuthFlow(AuthFlowType.REFRESH_TOKEN)
            .withClientId(appId)
            .withAuthParameters(refreshParams);

        authResult = identityProviderClient.initiateAuth(authRequest);
        System.out.println(authResult);
    }
}
```

**Comment by kink80** [comment]
I can see it now, you're using a device and that needs to be confirmed prior to using it in a refresh flow. If you had disabled use of devices for the user pool you won't need to use the device key.

Seems like I have a new use case to support here.

**Comment by henri-tremblay** [comment]
How do you see I am using a device?

**Comment by kink80** [comment]
It is in the user pool settings, it was turned off in my initial configuration.
If this was off I didn't have to perform
`String deviceKey = authenticationResultType.getNewDeviceMetadata().getDeviceKey();`
and pass it along.

**Comment by henri-tremblay** [comment]
I tried without passing the key along. It fails as well. But I've solved the mystery and indeed it's something we can add. The solution is to confirm the device. Which is possibly what you were talking about. It has to be done right after authentication.

I can't provide the code of `generateVerificationParameters`.

```java
String deviceKey = authenticationResultType.getNewDeviceMetadata().getDeviceKey();
String deviceGroup = authenticationResultType.getNewDeviceMetadata().getDeviceGroupKey();

Map<String, String> params = AWSCognitoSession.generateVerificationParameters(deviceKey, deviceGroup);

String passwordVerifier = params.get("verifier");
String salt = params.get("salt");

DeviceSecretVerifierConfigType deviceConfig = new DeviceSecretVerifierConfigType();
deviceConfig.setPasswordVerifier(passwordVerifier);
deviceConfig.setSalt(salt);

ConfirmDeviceRequest confirmDeviceRequest = new ConfirmDeviceRequest();
confirmDeviceRequest.setAccessToken(authenticationResultType.getAccessToken());  
confirmDeviceRequest.setDeviceKey(authenticationResultType.getNewDeviceMetadata().getDeviceKey());
confirmDeviceRequest.setDeviceName("someDeviceName");
confirmDeviceRequest.setDeviceSecretVerifierConfig(deviceConfig);
ConfirmDeviceResult confirmDeviceResult = identityProviderClient.confirmDevice(confirmDeviceRequest);
```

**Comment by kink80** [comment]
That's exactly what I had in my mind.

**Comment by kink80** [comment]
See the [PR](https://github.com/kink80/cognitoauth/pull/3) for implementation

**Comment by henri-tremblay** [comment]
Works perfectly. Fixed by #3
